### ISSUE TITLE
DESIGN: `iterator!` datatype

### ISSUE THREAD

**Issue by hiiamboris** [red/REP#168]
Read here: https://github.com/greggirwin/red-hof/blob/master/code-analysis/iterators.md

Based off my experience with the [extended `*each` loops](https://codeberg.org/hiiamboris/red-common/src/branch/master/foreach-design.md) and crafted to work together with the [`node!` datatype proposal](https://github.com/red/REP/issues/57), and `range!` (including series slices).

**Comment by hiiamboris** [comment]
A related (but OOP-like) design by @rgchris: https://gist.github.com/rgchris/9770ab0f7b625ed4b5db84a97fa999ba (also discussed [here](https://rebol.metaeducation.com/t/rgchriss-iterator-framework-in-oldes-rebol3/2564/4))

**Comment by hiiamboris** [comment]
A good coding challenge here that this design must be able to solve: https://rebol.metaeducation.com/t/for-both-loop-composability-for-the-win/1833

**Comment by greggirwin** [comment]
Read the post, and revisited many of your old notes as well, @hiiamboris. Thanks, once again, for all the deep diving and writing it down.

I'm all for unification, when we can, and when it makes sense. Also for making sure we figure out how to de-risk things, like `break/continue`. Hard to follow the Ren-C details without understanding all the other changes, types, and forms, so no thoughts on that here.

As you note near the end of one iterator doc, a key question is value versus complexity. Being able to combine things is very cool. When done outside a dialect, I think it can be more confusing in the grand scheme. e.g. `above/below` for graphs, and some of those use cases, are more niche than general. Not that we shouldn't support writing graph-based algos and such, but Red is high level, so that feels more like a dialect target to me.

As with `for` and general loop replacement, which you also note, there will be some adjustment, but also having meaningful words (`repeat, while, ....`) can cue you in to how code works, or at least how the author thought about it. After reading up on the Quorum language and their research, I leaned toward `repeat` as a better general name, but it also has a solid "count" model, unless we unify it and figure out the details we can all compromise on.

All this work is enormously valuable, especially analyzing code for patterns.

For me, it's correctness first, which means thinking about things like `break`, and no accidental endless loops. Then maybe see if doing the top 5 HOF or related improvements gives us good ROI for just a little work.
